tenacity/asyncio/retry.py

```diff
@@ -0,0 +1,283 @@
+# Copyright 2016–2021 Julien Danjou
+# Copyright 2016 Joshua Harlow
+# Copyright 2013-2014 Ray Holder
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import abc
+import re
+import typing
+
+from tenacity import retry_base
+
+if typing.TYPE_CHECKING:
+    from tenacity import RetryCallState
+
+
+class retry_base(retry_base):  # type: ignore[no-redef]
+    """Abstract base class for retry strategies."""
+
+    @abc.abstractmethod
+    async def __call__(self, retry_state: "RetryCallState") -> bool:  # type: ignore[override]
+        pass
+
+
+RetryBaseT = typing.Union[retry_base, typing.Callable[["RetryCallState"], typing.Awaitable[bool]]]
+
+
+class _retry_never(retry_base):
+    """Retry strategy that never rejects any result."""
+
+    async def __call__(self, retry_state: "RetryCallState") -> bool:  # type: ignore[override]
+        return False
+
+
+retry_never = _retry_never()
+
+
+class _retry_always(retry_base):
+    """Retry strategy that always rejects any result."""
+
+    async def __call__(self, retry_state: "RetryCallState") -> bool:  # type: ignore[override]
+        return True
```

Comment:
Is there a way for those function/classes to use the code from the sync ones? That'd avoid duplicating the logic everywhere.

Reply:
Good point. Besides, I also realised most of the implementations don't even need a redefinition as they are not performing any async operations. So I have removed most of the reimplementations and I've tried to DRY some of the others that use predicates, wdyt? https://github.com/hasier/tenacity/pull/2/commits/c619e1056741c7c1b74104488300d1e28fe6f72d

This implementation still looks somewhat odd to me though, I feel that combining sync and async retries may still yield some weird results, e.g. if the sync condition goes first in an `&`/`|` condition, then it will chain the rest as sync, which might not always work. Any suggestions to work around that? I was thinking of maybe referencing the async wrapper from the base class, but it sounds messy...

Reply:
I _think_ I found a solution for the issue. I have rebased this PR onto `main` and added the changes there, would be great to get your thoughts on that 🙂 https://github.com/jd/tenacity/pull/451